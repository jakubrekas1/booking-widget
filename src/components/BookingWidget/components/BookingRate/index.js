export { default as BookingRate } from './BookingRateContainer';
