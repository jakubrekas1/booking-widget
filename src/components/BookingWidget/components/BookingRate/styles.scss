@import '../../../../assets/styles/colors';

.booking-rate{
  display: flex;

  &__star{
    display: block;
    align-items: center;
    width: 8px;
    height: 16px;
    background-image: url('../../../../assets/images/star.svg');
    background-repeat: no-repeat;
    background-size: 16px;

    &--fill{
      background-image: url('../../../../assets/images/star-fill.svg');
    }

    &:nth-child(odd){
      background-position: left center;
    }

    &:nth-child(even){
      background-position: right center;
    }
  }

  &__votes{
    margin-left: 5px;
    color: $black;
    font-size: 14px;
    font-weight: bold;
  }
}
