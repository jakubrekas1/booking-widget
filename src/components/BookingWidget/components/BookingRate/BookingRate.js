import React from 'react';
import PropTypes from 'prop-types';

import './styles.scss';

const BookingRate = ({ stars, votes }) => (
  <div className="booking-rate">
    {stars}
    <div className="booking-rate__votes">{votes}</div>
  </div>
);

BookingRate.defaultProps = {
  votes: 0,
};

BookingRate.propTypes = {
  stars: PropTypes.arrayOf(PropTypes.node).isRequired,
  votes: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
};

export default BookingRate;
