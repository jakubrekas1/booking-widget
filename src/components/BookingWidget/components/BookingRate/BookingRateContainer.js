import React, { useMemo } from 'react';
import PropTypes from 'prop-types';

import { STARS_NUMBER } from '../../../../constants';
import BookingRate from './BookingRate';

const BookingRateContainer = ({ rate, votes }) => {
  const formattedRate = useMemo(() => Math.fround(rate) * 2, [rate]);

  const stars = useMemo(() => [...Array(STARS_NUMBER * 2)].map((star, index) => {
    const isFilled = index < formattedRate ? 'booking-rate__star--fill' : '';
    return (
      <div
        className={`booking-rate__star ${isFilled}`}
        // No unique values in the array
        // eslint-disable-next-line
        key={index}
      />
    );
  }), [formattedRate]);

  return (
    <BookingRate stars={stars} votes={votes} />
  );
};

BookingRateContainer.defaultProps = {
  rate: 0,
  votes: 0,
};

BookingRateContainer.propTypes = {
  rate: PropTypes.number,
  votes: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
};

export default BookingRateContainer;
