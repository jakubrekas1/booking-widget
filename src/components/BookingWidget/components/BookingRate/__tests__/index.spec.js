import React from 'react';
import renderer from 'react-test-renderer';
import BookingRate from '../BookingRateContainer';

describe('Booking rate component', () => {
  it('should renders correctly - rate 0', () => {
    const Component = <BookingRate rate={0} votes={0} />;
    const tree = renderer.create(Component).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should renders correctly with default props', () => {
    const Component = <BookingRate />;
    const tree = renderer.create(Component).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should renders correctly - rate 1', () => {
    const Component = <BookingRate rate={1} votes={20} />;
    const tree = renderer.create(Component).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should renders correctly - rate 1.5', () => {
    const Component = <BookingRate rate={1.5} votes={25} />;
    const tree = renderer.create(Component).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should renders correctly - rate 2', () => {
    const Component = <BookingRate rate={2} votes={50} />;
    const tree = renderer.create(Component).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should renders correctly - rate 4.5', () => {
    const Component = <BookingRate rate={4.5} votes={50} />;
    const tree = renderer.create(Component).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should renders correctly - rate 4.33', () => {
    const Component = <BookingRate rate={4.33} votes={50} />;
    const tree = renderer.create(Component).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should renders correctly - rate 5, all stars are filled', () => {
    const Component = <BookingRate rate={5} votes={50} />;
    const tree = renderer.create(Component).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
