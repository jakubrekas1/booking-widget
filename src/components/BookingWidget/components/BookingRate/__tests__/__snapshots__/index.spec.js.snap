// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Booking rate component should renders correctly - rate 0 1`] = `
<div
  className="booking-rate"
>
  <div
    className="booking-rate__star "
  />
  <div
    className="booking-rate__star "
  />
  <div
    className="booking-rate__star "
  />
  <div
    className="booking-rate__star "
  />
  <div
    className="booking-rate__star "
  />
  <div
    className="booking-rate__star "
  />
  <div
    className="booking-rate__star "
  />
  <div
    className="booking-rate__star "
  />
  <div
    className="booking-rate__star "
  />
  <div
    className="booking-rate__star "
  />
  <div
    className="booking-rate__votes"
  >
    0
  </div>
</div>
`;

exports[`Booking rate component should renders correctly - rate 1 1`] = `
<div
  className="booking-rate"
>
  <div
    className="booking-rate__star booking-rate__star--fill"
  />
  <div
    className="booking-rate__star booking-rate__star--fill"
  />
  <div
    className="booking-rate__star "
  />
  <div
    className="booking-rate__star "
  />
  <div
    className="booking-rate__star "
  />
  <div
    className="booking-rate__star "
  />
  <div
    className="booking-rate__star "
  />
  <div
    className="booking-rate__star "
  />
  <div
    className="booking-rate__star "
  />
  <div
    className="booking-rate__star "
  />
  <div
    className="booking-rate__votes"
  >
    20
  </div>
</div>
`;

exports[`Booking rate component should renders correctly - rate 1.5 1`] = `
<div
  className="booking-rate"
>
  <div
    className="booking-rate__star booking-rate__star--fill"
  />
  <div
    className="booking-rate__star booking-rate__star--fill"
  />
  <div
    className="booking-rate__star booking-rate__star--fill"
  />
  <div
    className="booking-rate__star "
  />
  <div
    className="booking-rate__star "
  />
  <div
    className="booking-rate__star "
  />
  <div
    className="booking-rate__star "
  />
  <div
    className="booking-rate__star "
  />
  <div
    className="booking-rate__star "
  />
  <div
    className="booking-rate__star "
  />
  <div
    className="booking-rate__votes"
  >
    25
  </div>
</div>
`;

exports[`Booking rate component should renders correctly - rate 2 1`] = `
<div
  className="booking-rate"
>
  <div
    className="booking-rate__star booking-rate__star--fill"
  />
  <div
    className="booking-rate__star booking-rate__star--fill"
  />
  <div
    className="booking-rate__star booking-rate__star--fill"
  />
  <div
    className="booking-rate__star booking-rate__star--fill"
  />
  <div
    className="booking-rate__star "
  />
  <div
    className="booking-rate__star "
  />
  <div
    className="booking-rate__star "
  />
  <div
    className="booking-rate__star "
  />
  <div
    className="booking-rate__star "
  />
  <div
    className="booking-rate__star "
  />
  <div
    className="booking-rate__votes"
  >
    50
  </div>
</div>
`;

exports[`Booking rate component should renders correctly - rate 4.5 1`] = `
<div
  className="booking-rate"
>
  <div
    className="booking-rate__star booking-rate__star--fill"
  />
  <div
    className="booking-rate__star booking-rate__star--fill"
  />
  <div
    className="booking-rate__star booking-rate__star--fill"
  />
  <div
    className="booking-rate__star booking-rate__star--fill"
  />
  <div
    className="booking-rate__star booking-rate__star--fill"
  />
  <div
    className="booking-rate__star booking-rate__star--fill"
  />
  <div
    className="booking-rate__star booking-rate__star--fill"
  />
  <div
    className="booking-rate__star booking-rate__star--fill"
  />
  <div
    className="booking-rate__star booking-rate__star--fill"
  />
  <div
    className="booking-rate__star "
  />
  <div
    className="booking-rate__votes"
  >
    50
  </div>
</div>
`;

exports[`Booking rate component should renders correctly - rate 4.33 1`] = `
<div
  className="booking-rate"
>
  <div
    className="booking-rate__star booking-rate__star--fill"
  />
  <div
    className="booking-rate__star booking-rate__star--fill"
  />
  <div
    className="booking-rate__star booking-rate__star--fill"
  />
  <div
    className="booking-rate__star booking-rate__star--fill"
  />
  <div
    className="booking-rate__star booking-rate__star--fill"
  />
  <div
    className="booking-rate__star booking-rate__star--fill"
  />
  <div
    className="booking-rate__star booking-rate__star--fill"
  />
  <div
    className="booking-rate__star booking-rate__star--fill"
  />
  <div
    className="booking-rate__star booking-rate__star--fill"
  />
  <div
    className="booking-rate__star "
  />
  <div
    className="booking-rate__votes"
  >
    50
  </div>
</div>
`;

exports[`Booking rate component should renders correctly - rate 5, all stars are filled 1`] = `
<div
  className="booking-rate"
>
  <div
    className="booking-rate__star booking-rate__star--fill"
  />
  <div
    className="booking-rate__star booking-rate__star--fill"
  />
  <div
    className="booking-rate__star booking-rate__star--fill"
  />
  <div
    className="booking-rate__star booking-rate__star--fill"
  />
  <div
    className="booking-rate__star booking-rate__star--fill"
  />
  <div
    className="booking-rate__star booking-rate__star--fill"
  />
  <div
    className="booking-rate__star booking-rate__star--fill"
  />
  <div
    className="booking-rate__star booking-rate__star--fill"
  />
  <div
    className="booking-rate__star booking-rate__star--fill"
  />
  <div
    className="booking-rate__star booking-rate__star--fill"
  />
  <div
    className="booking-rate__votes"
  >
    50
  </div>
</div>
`;

exports[`Booking rate component should renders correctly with default props 1`] = `
<div
  className="booking-rate"
>
  <div
    className="booking-rate__star "
  />
  <div
    className="booking-rate__star "
  />
  <div
    className="booking-rate__star "
  />
  <div
    className="booking-rate__star "
  />
  <div
    className="booking-rate__star "
  />
  <div
    className="booking-rate__star "
  />
  <div
    className="booking-rate__star "
  />
  <div
    className="booking-rate__star "
  />
  <div
    className="booking-rate__star "
  />
  <div
    className="booking-rate__star "
  />
  <div
    className="booking-rate__votes"
  >
    0
  </div>
</div>
`;
